packages/router/src/matcher/index.ts

```diff
@@ -277,8 +277,15 @@ export function createRouterMatcher(
         paramsFromLocation(
           currentLocation.params,
           // only keep params that exist in the resolved location
-          // TODO: only keep optional params coming from a parent record
-          matcher.keys.filter(k => !k.optional).map(k => k.name)
+          // only keep optional params coming from a parent record
+          matcher.keys
+            .filter(k => !k.optional)
+            .map(k => k.name)
```

Comment:
This could be done once at the end

Reply:
1.  A direct parent's keys contains all optional params, includes ancestors
2. I'll open a new PR to improve the code if the behavior it is necessary, absolutely i think this is better to keep the optional params from parent record.

```js
{
    path: '/:type?/root',
    children: [
    {
        path: 'path_a',
        name: 'a',
    },
    {
        path: 'path_b',
        name: 'b',
    }
  ]
}

// current path:   '/optionType/root/path_a'

resolve({
    name: 'b'
});
```
I think `/optionType/root/path_b` is better than `/root/path_b`  in this case

Reply:
I guess o forgot how I wrote my own code 😂

Reply:
I suggest to keep optional params coming from a parent record,  that is really useful